### PR TITLE
adding pathfinder2 services

### LINKED ISSUE #65: Integrate pathfinder2 services in docker setup
Create these services: blockchain-indexer, database, pathfinder-updater, pathfinder2 from  https://github.com/CirclesUBI/circles-platform/blob/main/ansible/roles/apps/templates/docker-compose.yaml.j2   in the docker setup.

### PR DESCRIPTION
**Pull request by JacqueGM**
- Add pathfinder2 services
 - Create `docker-compose.pathfinder-pull.yml`
 - Update `docker-compose.yml` with restart condition
 - Update Makefile to manage the pathfinder images too
 - Add other-blockchain-user to emulate activity in the network
 - Add more balance to eth accounts
 - Add pathfinder variables to `.env` file
 - Update README with command to start the pathfinder images
 - Update context indexer-db-init
 - Update seed origin path
 - Update migrate-contracts script to update the addresses
 - Use docker compose V2
 - Sync with new contributions in `land-local` repository
 - Add blockchain indexer
 - Add volumes for data in indexer and pathfinder
 - Use params for postgresql connections

Closes #65

### CODE REVIEW

File: docker-compose.pathfinder-pull.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  
+  init-db:
+    restart: "no"
+    depends_on:
+      - db
+    image: postgres:12
+    command: "sh -c 'psql postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}/${POSTGRES_DATABASE_INDEXER_DB} < /docker-entrypoint-initdb.d/0.sql'"
+    volumes:
+      - ./0.sql:/docker-entrypoint-initdb.d/0.sql  
+
```
Comment (almereyda) [inline docker-compose.pathfinder-pull.yml:45]: I would suggest to eliminate the whitespace below before merging.

File: docker-compose.yml [hunk 2]
```diff
@@ -10,7 +10,8 @@ volumes:
 services:
   db:
     image: postgres:12-alpine
-    container_name: circles-db
+    restart: unless-stopped
+    container_name: "${POSTGRES_HOST}"
     env_file:
```
Comment (almereyda) [inline docker-compose.yml:15]: Would it feel suitable to replace this `env_file:` declarative in this row with an explicit `environment:` dictionary that explicitly passes only those variables from `.env`, which are actually needed? This is also more verbose, and thus readable, understandable and explicit.

```suggestion
    environment:
```

File: docker-compose.yml [hunk 3]
```diff
@@ -10,7 +10,8 @@ volumes:
...
       - .env
```
Comment (almereyda) [inline docker-compose.yml:16]:
```suggestion
      POSTGRES_DB:       ${POSTGRES_USER}
      POSTGRES_USER:     ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
```

File: docker-compose.pathfinder-pull.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      INDEXER_CONNECTION_STRING: "Server=${POSTGRES_HOST};Port=5432;Database=${POSTGRES_DATABASE_INDEXER_DB};User ID=${POSTGRES_USER};Password=${POSTGRES_USER};Command Timeout=240"
+  
+  init-db:
+    restart: "no"
+    depends_on:
+      - db
+    image: postgres:12
+    command: "sh -c 'psql postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}/${POSTGRES_DATABASE_INDEXER_DB} < /docker-entrypoint-initdb.d/0.sql'"
+    volumes:
+      - ./0.sql:/docker-entrypoint-initdb.d/0.sql
```
[the remark this answers is not in the document]
Comment (almereyda) [reply docker-compose.pathfinder-pull.yml:44]:
```suggestion
      - ${PWD}/0.sql:/docker-entrypoint-initdb.d/0.sql
```

[the remark this answers is not in the document]
Comment (almereyda) [reply docker-compose.pathfinder-pull.yml:44]: Look what we've found! Some whitespace hid at the end of the line.

**Review by almereyda — CHANGES_REQUESTED** [review]
This came very well *together*, in the literate sense.

Let's please find a moment to consider the suggested code changes, and wait a bit until we can replace the image names with new permanent identifiers.

File: docker-compose.pathfinder-pull.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
+version: '3'
+services:
+  pathfinder2:
+    image: ghcr.io/circlesland/pathfinder2:1.4.0
```
[the remark this answers is not in the document]
Comment (almereyda) [reply docker-compose.pathfinder-pull.yml:4]: Please choose your flavour:

- https://hub.docker.com/r/circlesubi/pathfinder2/tags

```suggestion
    image: docker.io/circlesubi/pathfinder2:feature-dockerfile
```

- https://github.com/CirclesUBI/pathfinder2/pkgs/container/pathfinder2

```suggestion
    image: ghcr.io/circlesubi/pathfinder2:feature-dockerfile
```

File: docker-compose.pathfinder-pull.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      - "54389"
+    command:
+      - "0.0.0.0:54389"
+    volumes:
+      - ${PWD}/.state/pathfinder:/var/pathfinder2/data
+ 
+  pathfinder2-updater:
+    depends_on:
+      - pathfinder2
+    image: ghcr.io/circlesland/pathfinder2-updater:0.1.1
```
[the remark this answers is not in the document]
Comment (almereyda) [reply docker-compose.pathfinder-pull.yml:15]: Please choose your flavour:

- https://hub.docker.com/r/circlesubi/pathfinder2-updater/tags

```suggestion
    image: docker.io/circlesubi/pathfinder2-updater:dev
```

- https://github.com/CirclesUBI/pathfinder2/pkgs/container/pathfinder2

```suggestion
    image: ghcr.io/circlesubi/pathfinder2-updater:dev
```

File: docker-compose.pathfinder-pull.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      INDEXER_DB_CONNECTION_STRING: "Server=${POSTGRES_HOST};Port=5432;Database=${POSTGRES_DATABASE_INDEXER_DB};User ID=${POSTGRES_USER};Password=${POSTGRES_USER};Command Timeout=240"
+      INDEXER_WS_URL:               ${INDEXER_WS_URL}
+      INTERNAL_CAPACITY_GRAPH_PATH: ${INTERNAL_CAPACITY_GRAPH_PATH}
+      EXTERNAL_CAPACITY_GRAPH_PATH: ${EXTERNAL_CAPACITY_GRAPH_PATH}
+      PATHFINDER_RPC_URL:           ${PATHFINDER_RPC_URL}
+    volumes:
+      - ${PWD}/.state/pathfinder:/var/pathfinder2/data
+
+  blockchain-indexer:
+    image: ghcr.io/circlesland/blockchain-indexer:0.0.67
```
[the remark this answers is not in the document]
Comment (almereyda) [reply docker-compose.pathfinder-pull.yml:27]: Please choose your flavour:

- https://hub.docker.com/r/circlesubi/blockchain-indexer/tags

```suggestion
    image: docker.io/circlesubi/blockchain-indexer:dev
```

- https://github.com/CirclesUBI/blockchain-indexer/pkgs/container/blockchain-indexer

```suggestion
    image: ghcr.io/circlesubi/blockchain-indexer:dev
```

**Comment by almereyda** [comment]
@jaensen also wanted to point your attention at the fact that there now is also a:

- https://github.com/CirclesUBI/pathfinder-proxy
  - https://github.com/CirclesUBI/pathfinder-proxy/pkgs/container/pathfinder-proxy
  - https://hub.docker.com/r/circlesubi/pathfinder-proxy/tags

I am sure they would be open to explain and introduce it in this round, in so this new Pathfinder setup is near currently possible feature completeness.

**Comment by llunaCreixent** [comment]
Task: add the pathfinder-proxy for test it: https://hub.docker.com/orgs/circlesubi/repositories
https://github.com/CirclesUBI/pathfinder-proxy

File: docker-compose.pathfinder-pull.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
+version: '3'
+services:
+  pathfinder2:
```
[the remark this answers is not in the document]
Comment (JacqueGM) [reply docker-compose.pathfinder-pull.yml:3]: added

**Comment by llunaCreixent** [comment]
Task:
- [ ] Add the env vars:

```
me says:WARNING: The POSTGRES_HOST variable is not set. Defaulting to a blank string.
WARNING: The POSTGRES_DATABASE_INDEXER_DB variable is not set. Defaulting to a blank string.
WARNING: The INDEXER_WS_URL variable is not set. Defaulting to a blank string.
WARNING: The INTERNAL_CAPACITY_GRAPH_PATH variable is not set. Defaulting to a blank string.
WARNING: The EXTERNAL_CAPACITY_GRAPH_PATH variable is not set. Defaulting to a blank string.
WARNING: The PATHFINDER_RPC_URL variable is not set. Defaulting to a blank string.
WARNING: The INDEXER_WEBSOCKET_URL variable is not set. Defaulting to a blank string.
WARNING: The INDEXER_RPC_GATEWAY_URL variable is not set. Defaulting to a blank string. 
```

**Comment by llunaCreixent** [comment]
Task:
- [ ] add the `other-blockchain-user` like in here: https://github.com/CirclesUBI/land-local/blob/main/docker-compose.yml

So that the `pathfinder-updater` doesn't assume the system is unhealthy

File: README.md [hunk 9]
```diff
@@ -62,6 +62,13 @@ Enables you to start Circles services in Docker containers for Circles developme
     127.0.1.1 graph.circles.local
     127.0.1.1 relay.circles.local
     ```
+  You can run the following commands to do this automatically (on linux):
+  ```bash
+  # Add entries to /etc/hosts if they don't exist
+  grep -q "^127.0.0.1 api.circles.lan$" /etc/hosts || echo "127.0.0.1 api.circles.lan" | sudo tee -a /etc/hosts
+  grep -q "^127.0.0.1 graph.circles.lan$" /etc/hosts || echo "127.0.0.1 graph.circles.lan" | sudo tee -a /etc/hosts
+  grep -q "^127.0.0.1 relay.circles.lan$" /etc/hosts || echo "127.0.0.1 relay.circles.lan" | sudo tee -a /etc/hosts
```
Comment (llunaCreixent) [inline README.md:70]: Here it should say `circles.local`, right?

File: README.md [hunk 10]
```diff
@@ -145,6 +152,31 @@ For preparing the environment to run the [`circles-core` tests](https://github.c
...
+
+```
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml up indexer-db (wait until db ready)
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml up -d indexer-db
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml run --rm indexer-db-init 
+```
+After this launch the rest of the services as usual
+
+```
+make up API EXPOSE_PORTS=1 && make contracts && make subgraph && make up   EXPOSE_PORTS=1
```
Comment (llunaCreixent) [inline README.md:176]: Does the command doesn't need the `API` arg?

**Comment by llunaCreixent** [comment]
when running the new docker compose, with my debian system, I get:
```
PermissionError: [Errno 13] Permission denied: '/home/CirclesUBI/circles-docker/.state/indexer-db/pg_data'
make: *** [Makefile:63: up] Error 1
```
```
$ ls -la /home/CirclesUBI/circles-docker/.state/indexer-db/
total 12
drwxr-xr-x  3 root             root 4096 Feb 14 15:35 .
drwxr-xr-x  4 root             root 4096 Feb 14 15:35 ..
drwx------ 19 systemd-coredump root 4096 Feb 14 16:03 pg_data
```

**Comment by almereyda** [comment]
In https://github.com/CirclesUBI/circles-docker/commit/88de80c7563a5b0395b878986c37e693120b3323 this is adding instructions to use a priviledged account to get a working development environment. This would be highly discoured, from a security point of view. An alternative could be to run a DNS resolver for the `circles.lan` zone that is introduced, or to use subdomains from localhost.

In development of `land-local` we were very surprised when we found out that we could simply use subdomains from localhost to address individual services proxied behind the same endpoint.

Please feel free to take inspiration in https://github.com/CirclesUBI/land-local/blob/dev/modules/caddy/Caddyfile (excluding TLS handling) and consider if that could be adapted to your Docker environment: `*.circles.localhost` doesn't sound too bad, either.

Taking care of this would bring less lines of code that manage state outside of the container (here: `/etc/hosts`), and bring increased security by not needing developers to change core system components as root.

Also it may appear as a break in the levels of abstraction, when a containerised environment requires changes to its host system outside of the containers.

File: README.md [hunk 11]
```diff
@@ -145,6 +152,42 @@ For preparing the environment to run the [`circles-core` tests](https://github.c
...
+3.  
+make pathfinder
+
+4. 
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml up indexer-db # (wait until db ready)
+
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml up -d indexer-db
+
+docker compose -p  circles -f docker-compose.pathfinder-pull.yml run --rm indexer-db-init 
+```
```
Comment (llunaCreixent) [inline README.md:186]: that is actually not needed

**Review by llunaCreixent — DISMISSED** [review]
All goooood!!
Core tests (integration) pass with pathfinder server type.

**Review by llunaCreixent — APPROVED** [review]
All tests patthing with both cli and server pathfinderType

**Comment by llunaCreixent** [comment]
@almereyda can you review this PR? It seems there's a requested change from you blocking the merge.
Thanks! :)

**Review by juanenrisley — APPROVED** [review]
👍🏿

**Pull request merged**